prac_07/project_management.py

```diff
@@ -0,0 +1,160 @@
+        elif choice == "F":
+            filter_projects(projects)
+        elif choice == "A":
+            add_project(projects)
+        elif choice == "U":
+            update_project(projects)
+        else:
+            print("Invalid menu choice")
+        choice = input(MENU).upper()
+    print("Thank you for using custom-built project management software.")
+
+
+def load_projects(filename):
+    """Load projects from user file and store in a nested list."""
+    projects = []
+    with open(filename) as in_file:
+        in_file.readline()
+        for line in in_file:
+            parts = line.strip().split("\t")
+            start_date = datetime.datetime.strptime(parts[1], "%d/%m/%Y").date()
+            new_project = Project(parts[0], start_date, int(parts[2]), float(parts[3]), parts[4])
+            projects.append(new_project)
+    return projects
+
+
+def save_project(projects):
+    file_to_save = input("Filename: ")
+    with open(file_to_save, "w") as out_file:
+        for project in projects:
+            print(project, file=out_file)
+
+
+def display_projects(incomplete_projects, complete_projects):
+    """Display formatted projects."""
+    print(f"Incomplete projects:")
+    for project in incomplete_projects:
+        print(f"\t {project}")
+    print(f"Complete projects:")
+    for project in complete_projects:
+        print(f"\t {project}")
+
+
+def filter_projects(projects):
+    """Filter projects by date and display sorted projects that start after that date."""
+    filter_date_string = input("Show projects that start after date (dd/mm/yy): ")
+    filter_date = datetime.datetime.strptime(filter_date_string, "%d/%m/%Y").date()
+    filtered_projects = []
+    for project in projects:
+        if project.start_date >= filter_date:
+            filtered_projects.append(project)
+    filtered_projects.sort()
+    for project in filtered_projects:
+        print(project)
+
+
+def add_project(projects):
+    """Add new project to projects after error-checking user input."""
+    print("Let's add a new project")
+    name = input("Name: ")
+    start_date_string = input("Start date (dd/mm/yy): ")
+    start_date = datetime.datetime.strptime(start_date_string, "%d/%m/%Y").date()
+    priority = int(input("Priority: "))
+    cost = float(input("Cost estimate: $"))
+    completion_percentage = input("Percent complete: ")
+    new_project = Project(name, start_date, priority, cost, completion_percentage)
+    projects.append(new_project)
+
+
+def update_project(projects):
+    """Prompt user to update a project completion percentage and priority by number (error-checked)."""
+    for i, project in enumerate(projects):
+        print(f"{i} {project}")
+    is_valid_project_index = False
+    while not is_valid_project_index:
+        try:
+            project_to_update_index = int(input("Project choice: "))
+            if project_to_update_index < 1:
+                print("Number must be >= 1")
```

Comment:
I can't enter the first project number (index 0) because of your error checking. Be careful with your indexing and error checking 😃